source/common/stats/thread_local_store.h

```diff
@@ -16,6 +16,142 @@
 namespace Envoy {
 namespace Stats {
 
+/**
+ * Log Linear Histogram implementation per thread.
+ */
+class ThreadLocalHistogramImpl : public Histogram, public MetricImpl {
+public:
+  ThreadLocalHistogramImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                           std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent),
+        current_active_(0), flags_(0) {
+    histograms_[0] = hist_alloc();
+    histograms_[1] = hist_alloc();
+    cumulative_statistics_.reset(new HistogramStatisticsImpl());
+    interval_statistics_.reset(new HistogramStatisticsImpl());
+  }
+
+  ~ThreadLocalHistogramImpl() {
+    hist_free(histograms_[0]);
+    hist_free(histograms_[1]);
+  }
+  // Stats::Histogram
+  void recordValue(uint64_t value) override {
+    std::unique_lock<std::mutex> lock(merge_lock_);
```

Comment:
here lock may not be required. need to think through a bit here.

Reply:
Lock is not needed here. Because of the RCU locking style we are using it's always safe to directly hit the current histogram.